File: RestaurantSystem.Api/Features/Orders/Dtos/OrderDto.cs
namespace RestaurantSystem.Api.Features.Orders.Dtos;

public record OrderDto
{
    public Guid Id { get; set; }
    public string OrderNumber { get; set; } = null!;
    public Guid? UserId { get; set; }
    public string? CustomerName { get; set; }
    public string? CustomerEmail { get; set; }
    public string? CustomerPhone { get; set; }

    // Order Type
    public string Type { get; set; } = null!;
    public int? TableNumber { get; set; }

    // Pricing
    public decimal SubTotal { get; set; }
    public decimal Tax { get; set; }
    public decimal DeliveryFee { get; set; }
    public decimal Discount { get; set; }
    public decimal DiscountPercentage { get; set; }
    public decimal CustomerDiscountAmount { get; set; }
    public decimal Tip { get; set; }
    public decimal Total { get; set; }

    // Payment Summary
    public decimal TotalPaid { get; set; }
    public decimal RemainingAmount { get; set; }
    public bool IsFullyPaid { get; set; }

    // Status
    public string Status { get; set; } = null!;
    public string PaymentStatus { get; set; } = null!;

    // Focus Order
    public bool IsFocusOrder { get; set; }
    public int? Priority { get; set; }
    public string? FocusReason { get; set; }
    public DateTime? FocusedAt { get; set; }
    public string? FocusedBy { get; set; }

    // Timestamps
    public DateTime OrderDate { get; set; }
    public DateTime? EstimatedDeliveryTime { get; set; }
    public DateTime? ActualDeliveryTime { get; set; }
    public DateTime? CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }



    // Additional Info
    public string? Notes { get; set; }
    public DeliveryAddressDto? DeliveryAddress { get; set; }
    public string? CancellationReason { get; set; }

    public string? PromoCode { get; set; }
    public bool HasUserLimitDiscount { get; set; }
    public decimal UserLimitAmount { get; set; } // Threshold for discount

    // Related Data
    public List<OrderItemDto> Items { get; set; } = new();
    public List<OrderPaymentDto> Payments { get; set; } = new();
    public List<OrderStatusHistoryDto> StatusHistory { get; set; } = new();

}

File: src/types/order/orderDto.ts
/**
 * The aggregate order request/response shapes and their API-response aliases.
 * Extracted from types/order.ts (Sprint 4/6 type-file split by domain).
 */

import { OrderType } from './enums';
import { ApiResponse, PagedResult } from './common';
import {
  CreateOrderDeliveryAddressDto,
  DeliveryAddressDto,
  CreateOrderItemDto,
  OrderItemDto,
  CreateOrderPaymentDto,
  OrderPaymentDto,
  OrderStatusHistoryDto,
} from './dtos';
import type { OrderPermittedActionDto } from './tableBillRound';
import type { OrderRoutingStateDto } from './orderRouting';

/**
 * Create order command (request)
 */
export interface CreateOrderCommand {
  sessionId?: string;
  userId?: string;
  customerName?: string;
  customerEmail?: string;
  customerPhone?: string;
  type: OrderType;
  tableNumber?: number | null;
  promoCode?: string;
  hasUserLimitDiscount?: boolean;
  userLimitAmount?: number;
  // Pre-calculated values from basket (optional)
  basketSubTotal?: number;
  basketTax?: number;
  basketDiscount?: number;
  basketCustomerDiscount?: number;
  basketTotal?: number;
  // Fidelity Points
  pointsToRedeem?: number;
  // Tip
  tip?: number;
  isFocusOrder?: boolean;
  priority?: number;
  focusReason?: string;
  notes?: string;
  deliveryAddress?: CreateOrderDeliveryAddressDto;
  items?: CreateOrderItemDto[];
  payments?: CreateOrderPaymentDto[];
}

/**
 * Body for POST /api/Orders/from-basket (menu-bundles redesign #157, slice 5). The server reads the
 * user's persisted basket (resolved from the X-Session-Id header, added by apiClient) and derives the
 * order items itself — so, unlike CreateOrderCommand, this carries no `items`, and drops the
 * server/session-owned or staff/POS-only fields (`sessionId`, `userId`, focus-order, user-limit
 * discount).
 */
export type CreateOrderFromBasketCommand = Omit<
  CreateOrderCommand,
  | 'items'
  | 'sessionId'
  | 'userId'
  | 'isFocusOrder'
  | 'priority'
  | 'focusReason'
  | 'hasUserLimitDiscount'
  | 'userLimitAmount'
>;

/**
 * Complete order data (response)
 */
export interface OrderDto {
  id: string;
  orderNumber: string;
  userId?: string;
  customerName?: string;
  customerEmail?: string;
  customerPhone?: string;
  type: string;
  tableNumber?: number | null;
  /** Stable table identity for dine-in orders (additive S2 response field). */
  tableId?: string | null;
  /** Display label for a table, including labels that are not numeric (for example, T-QA). */
  tableLabel?: string | null;
  /** Explicit visit membership for staff-created dine-in rounds. */
  serviceSessionId?: string | null;
  subTotal: number;
  tax: number;
  deliveryFee: number;
  discount: number;
  discountPercentage: number;
  customerDiscountAmount: number;
  tip: number;
  total: number;
  totalPaid: number;
  remainingAmount: number;
  isFullyPaid: boolean;
  status: string;
  paymentStatus: string;
  /** Whether the order was deliberately released to the kitchen. */
  isKitchenReleased?: boolean;
  kitchenReleasedAt?: string | null;
  kitchenReleasedBy?: string | null;
  /** Server-issued aggregate version used for conditional staff mutations. */
  version: number;
  isFocusOrder: boolean;
  priority?: number;
  focusReason?: string;
  focusedAt?: string;
  focusedBy?: string;
  orderTypeOverrideBy?: string | null;
  orderTypeOverrideItems?: string | null;
  preferredLanguage?: string | null;
  orderDate: string;
  estimatedDeliveryTime?: string;
  actualDeliveryTime?: string;
  createdAt?: string;
  updatedAt?: string;
  notes?: string;
  deliveryAddress?: DeliveryAddressDto;
  cancellationReason?: string;
  promoCode?: string;
  hasUserLimitDiscount: boolean;
  userLimitAmount: number;
  currency?: string | null;
  items: OrderItemDto[];
  payments: OrderPaymentDto[];
  statusHistory: OrderStatusHistoryDto[];
  permittedActions?: OrderPermittedActionDto[] | null;
  routingStates?: OrderRoutingStateDto[] | null;
  /**
   * Per-order guest token (plan S3): lets the guest WATCH their own order via the anonymous
   * guest-status endpoint without holding any staff credential. Creation response only.
   */
  guestStatusToken?: string;
}

/**
 * Order API response types
 */
export type OrderDtoApiResponse = ApiResponse<OrderDto>;
export type OrderDtoListApiResponse = ApiResponse<OrderDto[]>;
export type OrderDtoPagedResultApiResponse = ApiResponse<PagedResult<OrderDto>>;
export type OrderPaymentDtoApiResponse = ApiResponse<OrderPaymentDto>;
